pkg/storegateway/series_refs.go

```diff
@@ -798,11 +878,135 @@ func (s *loadingSeriesChunkRefsSetIterator) Next() bool {
 
 	s.currentSet = nextSet
 	if s.skipChunks && cachedSeriesID.isSet() {
-		storeCachedSeriesForPostings(s.ctx, s.indexCache, s.tenantID, s.blockID, s.shard, cachedSeriesID, nextSet, s.logger)
+		storeCachedSeriesForPostings(ctx, s.indexCache, s.tenantID, s.blockID, s.shard, cachedSeriesID, nextSet, s.logger)
 	}
 	return true
 }
 
+// clampLastChunkLength checks the length of the last chunk in the last range of the last series.
+// If the length of that chunk is larger than the difference with the first chunk ref in metas
+// then the length is clamped at that difference.
+// clampLastChunkLength assumes that the chunks are sorted by their refs
+// (currently this is equivalent to also being sorted by their minTime)
+// clampLastChunkLength is a noop if metas or series is empty.
+func clampLastChunkLength(series []seriesChunkRefs, metas []chunks.Meta) {
+	if len(series) == 0 || len(metas) == 0 {
+		return
+	}
+	var (
+		lastSeriesRanges = series[len(series)-1].chunksRanges
+		lastRange        = lastSeriesRanges[len(lastSeriesRanges)-1]
+		lastSeriesChunk  = lastRange.refs[len(lastRange.refs)-1]
+		firstRef         = metas[0].Ref
+	)
+
+	if lastRange.segmentFile != uint32(chunkSegmentFile(firstRef)) {
+		return
+	}
+	diffWithNextChunk := chunkOffset(firstRef) - lastSeriesChunk.segFileOffset
+	if lastSeriesChunk.length > diffWithNextChunk {
+		lastRange.refs[len(lastRange.refs)-1].length = diffWithNextChunk
+	}
+}
+
+// partitionChunks creates a slice of seriesChunkRefsRange for each range of chunks within the same segment file.
+// It may also create more ranges if there are more chunks.
+// The partitioning here should be fairly static and not depend on the actual Series() request because
+// the resulting ranges may be used for caching, and we want our cache entries to be reusable between requests.
+func partitionChunks(chks []chunks.Meta) [][]chunks.Meta {
+	const (
+		chunksRangesPerSeries = 2
+		minChunksPerRange     = 2
+	)
+	chunksPerRange := util_math.Max(minChunksPerRange, len(chks)/chunksRangesPerSeries)
+
+	ranges := make([][]chunks.Meta, 0, util_math.Min(chunksRangesPerSeries, len(chks)/chunksPerRange))
```

Comment:
Shouldn't be +1? 🤔 

```suggestion
	ranges := make([][]chunks.Meta, 0, util_math.Min(chunksRangesPerSeries, len(chks)/chunksPerRange)+1)
```

Reply:
I'm not sure i understand why. We want either the target `chunksRangesPerSeries` number of ranges or however many times `minChunksPerRange` fits into `len(chks)`. 

If there is some remainder from `len(chks) / minChunksPerRange` we put all of it into the last range.

Reply:
You're right. Ignore my comment.